### PR TITLE
feat: add undefined keyword

### PR DESCRIPTION
**Pull request by mahzoun99**

### CODE REVIEW

File: shared/src/main/scala/mlscript/JSBackend.scala [hunk 1]
```diff
@@ -185,6 +185,7 @@ class JSBackend {
     case IntLit(value) => JSLit(value.toString + (if (JSBackend isSafeInteger value) "" else "n"))
     case DecLit(value) => JSLit(value.toString)
     case StrLit(value) => JSExpr(value)
+    case UnitLit(value) => if(value){JSExpr("undefined")} else{JSExpr("null")}
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/JSBackend.scala]: This actually creates a JS string literal, which is not what we want. I believe what we want is `JSLit(if (value) "undefined" else "null")`.

File: shared/src/main/scala/mlscript/Typer.scala [hunk 2]
```diff
@@ -442,6 +443,7 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool)
         // currently we get things like "flows into variable reference"
         // but we used to get the better "flows into object receiver" or "flows into applied expression"...
       case lit: Lit => ClassTag(lit, lit.baseClasses)(prov)
+      // HERE
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/Typer.scala]: Here?

File: shared/src/main/scala/mlscript/helpers.scala [hunk 3]
```diff
@@ -72,6 +72,7 @@ abstract class TypeImpl extends Located { self: Type =>
     case Literal(IntLit(n)) => n.toString
     case Literal(DecLit(n)) => n.toString
     case Literal(StrLit(s)) => "\"" + s + "\""
+    case Literal(UnitLit(b)) => if(b){"undefined"} else {"null"}
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/helpers.scala]: Formatting is non-standard.

```suggestion
    case Literal(UnitLit(b)) => if (b) "undefined" else "null"
```

File: shared/src/main/scala/mlscript/helpers.scala [hunk 4]
```diff
@@ -258,6 +260,7 @@ trait TermImpl extends StatementImpl { self: Term =>
     case IntLit(value) => value.toString
     case DecLit(value) => value.toString
     case StrLit(value) => '"'.toString + value + '"'
+    case UnitLit(value) => if(value){"undefined"} else {"null"}
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/helpers.scala]: Formatting.

File: shared/src/main/scala/mlscript/syntax.scala [hunk 5]
```diff
@@ -74,6 +74,7 @@ final case object NoCases extends CaseBranches
 final case class IntLit(value: BigInt)      extends Lit
 final case class DecLit(value: BigDecimal)  extends Lit
 final case class StrLit(value: Str)         extends Lit
+final case class UnitLit(undefinedOrNull: Bool)    extends Lit
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/syntax.scala]: Let's align all the `extends` clauses :^)

**Review by LPTK — COMMENTED** [review]
Cool, thanks!

Please address the comments I made on the PR.

Also, from some test output changes, it looks like you have an older (or newer?) version of nodejs. Can you tell me what `node -v` returns?

Finally, this needs a few more tests, and the tests should be moved to a different file – call it `Undefiend.mls`. For instance, let's add tests like:

```scala

case ()[0] of { undefined -> 1 }
//│ res: 1
//│    = 1

case ()[0] of { "" -> 0 | undefined -> 1 }
//│ res: 0 | 1
//│    = 1

case ()[0] of { null -> 0 | undefined -> 1 }
//│ ╔══[ERROR] type identifier not found: null
//│ ║  l.67: 	case ()[0] of { null -> 0 | undefined -> 1 }
//│ ╙──      	                ^^^^
//│ res: error
//│ Code generation encountered an error:
//│   unknown match case: null
```

(It would be good if the latter would also compile.)

File: shared/src/main/scala/mlscript/Typer.scala [hunk 6]
```diff
@@ -106,6 +106,7 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool)
     TypeDef(Cls, TypeName("true"), Nil, Nil, TopType, Nil, Nil, Set.single(Var("bool")), N) ::
     TypeDef(Cls, TypeName("false"), Nil, Nil, TopType, Nil, Nil, Set.single(Var("bool")), N) ::
     TypeDef(Cls, TypeName("string"), Nil, Nil, TopType, Nil, Nil, Set.empty, N) ::
+    TypeDef(Cls, TypeName("undefined"), Nil, Nil, TopType, Nil, Nil, Set.empty, N) ::
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/Typer.scala]: Actually this makes `undefined` a normal nominal class, but it's not – it's a literal class. It should be:
```suggestion
    TypeDef(Als, TypeName("undefined"), Nil, Nil, ClassTag(UnitLit(true), Set.empty)(noProv), Nil, Nil, Set.empty, N) ::
```

File: shared/src/main/scala/mlscript/helpers.scala [hunk 7]
```diff
@@ -318,6 +321,7 @@ trait LitImpl { self: Lit =>
     case _: IntLit => Set.single(Var("int")) + Var("number")
     case _: StrLit => Set.single(Var("string"))
     case _: DecLit => Set.single(Var("number"))
+    case _: UnitLit => Set.single(Var("undefined"))
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/helpers.scala]: `undefined` is not a parent type of `undefined`!

```suggestion
    case _: UnitLit => Set.empty
```

**Review by LPTK — CHANGES_REQUESTED** [review]
I found some more serious problems (see new comments below).

File: shared/src/main/scala/mlscript/Typer.scala (hunk 2 above)
Comment (mahzoun99) [reply shared/src/main/scala/mlscript/Typer.scala]: sorry, I was about to remove it after your review :sweat_smile:

**Comment by LPTK** [comment]
Great! Can you add a few tests? The ones I suggested are by no means sufficient. You should at least have tests for writing and using some basic `undefined` and `null` expressions!

**Pull request closed without merging**